src/plan.rs

```diff
@@ -0,0 +1,504 @@
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//! A spending plan or *plan* for short is a representation of a particular spending path on a
+//! descriptor. This allows us to analayze a choice of spending path without producing any
+//! signatures or other witness data for it.
+//!
+//! To make a plan you provide the descriptor with "assets" like which keys you are able to use, hash
+//! pre-images you have access to, absolute/relative timelock constraints etc.
+//!
+//! Once you've got a plan it can tell you its expected satisfaction weight which can be useful for
+//! doing coin selection. Furthermore it provides which subset of those keys and hash pre-images you
+//! will actually need as well as what locktime or sequence number you need to set.
+//!
+//! Once you've obstained signatures, hash pre-images etc required by the plan, it can create a
+//! witness/script_sig for the input.
+
+use core::cmp::Ordering;
+use core::iter::FromIterator;
+
+use bitcoin::hashes::{hash160, ripemd160, sha256};
+use bitcoin::util::address::WitnessVersion;
+use bitcoin::util::taproot::TapLeafHash;
+use bitcoin::{LockTime, Sequence};
+
+use crate::descriptor::{DescriptorType, KeyMap};
+use crate::miniscript::context::SigType;
+use crate::miniscript::hash256;
+use crate::miniscript::satisfy::{Placeholder, Satisfier, WitnessTemplate};
+use crate::prelude::*;
+use crate::util::witness_size;
+use crate::{
+    DefiniteDescriptorKey, DescriptorPublicKey, MiniscriptKey, ScriptContext, ToPublicKey,
+};
+
+/// Trait describing a present/missing lookup table for constructing witness templates
+///
+/// This trait mirrors the [`Satisfier`] trait with the difference that instad of returning the
+/// item if it's present, it only returns a boolean to indicate its presence.
+///
+/// This trait is automatically implemented for every type that is also a satisfier, and simply
+/// proxies the queries to the satisfier and returns whether an item is available or not.
+///
+/// All the methods have a default implementation that returns `false`.
+pub trait AssetProvider<Pk: MiniscriptKey> {
+    /// Given a public key, look up an ECDSA signature with that key
+    fn lookup_ecdsa_sig(&self, _: &Pk) -> bool {
+        false
+    }
+
+    /// Lookup the tap key spend sig
+    fn lookup_tap_key_spend_sig(&self, _: &Pk) -> bool {
+        false
+    }
+
+    /// Given a public key and a associated leaf hash, look up an schnorr signature with that key
+    fn lookup_tap_leaf_script_sig(&self, _: &Pk, _: &TapLeafHash) -> bool {
+        false
+    }
+
+    /// Given a raw `Pkh`, lookup corresponding `Pk`. If present, return its lenght.
+    fn lookup_raw_pkh_pk<Ctx: ScriptContext>(&self, _: &hash160::Hash) -> Option<usize> {
+        None
+    }
+
+    /// Given a raw `Pkh`, lookup corresponding [`bitcoin::XOnlyPublicKey`].
+    fn lookup_raw_pkh_x_only_pk(&self, _: &hash160::Hash) -> bool {
+        false
+    }
+
```

Comment:
We are dropping a reference to control block in here. which in satisfier is given as 
```rust
  /// Obtain a reference to the control block for a ver and script
    fn lookup_tap_control_block_map(
        &self,
    ) -> Option<&BTreeMap<ControlBlock, (bitcoin::Script, LeafVersion)>> {
        None
    }
```

Obvious reason would be that we don't need this into the Plan module but what it does ? Why it exist in the satisfier and not in plan. It has anything to do with the merkle root ?